File: analyzer/core/crashReportParser.js
const kFrameworkDlls = new Set([
  'redscope.dll', 'red4ext.dll', 'cyber_engine_tweaks.asi', 'scc.dll',
]);

const kSystemDlls = new Set([
  'ntdll.dll', 'kernel32.dll', 'kernelbase.dll', 'user32.dll', 'ucrtbase.dll',
  'msvcp140.dll', 'vcruntime140.dll', 'vcruntime140_1.dll',
  'gdi32.dll', 'gdi32full.dll', 'combase.dll', 'ole32.dll', 'oleaut32.dll',
  'shell32.dll', 'advapi32.dll', 'ws2_32.dll', 'd3d12.dll', 'dxgi.dll',
  'd3d11.dll', 'dbghelp.dll', 'msvcrt.dll', 'rpcrt4.dll', 'sechost.dll',
  'bcrypt.dll', 'bcryptprimitives.dll', 'win32u.dll',
]);

function classifyStackKind(name) {
  if (!name) return 'unknown';
  const n = name.toLowerCase();
  if (n === 'cyberpunk2077.exe') return 'game';
  if (kFrameworkDlls.has(n)) return 'framework';
  if (kSystemDlls.has(n)) return 'system';
  return 'mod';
}

function trim(s) { return (s == null ? '' : String(s)).replace(/^\s+|\s+$/g, ''); }
function startsWith(s, p) { return s.slice(0, p.length) === p; }

function splitLines(text) {
  if (!text) return [];
  return text.split(/\r?\n/);
}

function sectionLabel(line) {
  const m = line.match(/^---\s+(.*?)\s+-+\s*$/);
  return m ? m[1] : null;
}

function splitSections(lines) {
  const sections = { header: [] };
  const order = ['header'];
  let current = 'header';
  for (const line of lines) {
    const label = sectionLabel(line);
    if (label != null) {
      current = label;
      if (!sections[current]) { sections[current] = []; order.push(current); }
    } else {
      sections[current].push(line);
    }
  }
  return { sections, order };
}

function parseHeader(lines, out) {
  for (const line of lines) {
    const culprit = line.match(/^LIKELY CULPRIT:\s*(.*)$/);
    if (culprit) out.likelyCulprit = trim(culprit[1]);

    const idLine = line.match(/^CRASH ID:\s*(\S+)\s*\(loose\s+(\S+)\)/);
    if (idLine) { out.crashId = trim(idLine[1]); out.looseId = trim(idLine[2]); }

    const fp = line.match(/fingerprint\s+([0-9A-Fa-f]+)\s*\/\s*([0-9A-Fa-f]+)\s+modset\s+([0-9A-Fa-f]+)/);
    if (fp) out.fingerprint = { primary: fp[1], loose: fp[2], modSet: fp[3] };
  }
}

function parseCrashSection(lines, out) {
  for (const line of lines) {
    const exc = line.match(/^Exception\s*:\s*(0x[0-9A-Fa-f]+)\s*(.*)$/);
    if (exc) { out.exception = out.exception || {}; out.exception.code = exc[1]; out.exception.name = trim(exc[2]); }

    const addr = line.match(/^Address\s*:\s*(0x[0-9A-Fa-f]+)/);
    if (addr) { out.exception = out.exception || {}; out.exception.address = addr[1]; }

    const fault = line.match(/^Fault data:\s+(\S+)\s+at\s+(0x[0-9A-Fa-f]+)/);
    if (fault) out.faultData = { op: fault[1], address: fault[2] };

    const mod = line.match(/^\s+module:\s+(\S+)\s*\+\s*(0x[0-9A-Fa-f]+)/);
    if (mod) out.faultingModule = { name: mod[1], rva: mod[2] };

    const timeStr = line.match(/^Time\s*:\s*(.*)$/);
    if (timeStr) out.timeText = trim(timeStr[1]);

    const cls = line.match(/^Class\s*:\s*(.*)$/);
    if (cls) out.crashClass = trim(cls[1]);

    const rsv = line.match(/^REDscope\s*:\s*(.*)$/);
    if (rsv) out.redscopeVersion = trim(rsv[1]);

    const cp = line.match(/^Cyberpunk\s*:\s*(.*)$/);
    if (cp) out.buildId = trim(cp[1]);
  }
}

function parseNativeCallstack(lines, out) {
  for (const line of lines) {
    const m = line.match(/^\[\s*0\]\s+(0x[0-9A-Fa-f]+)\s+(\S+?)!.*?\+(0x[0-9A-Fa-f]+)\s+<faulting RIP>/);
    if (m) {
      out.faultingModule = out.faultingModule || { name: m[2], rva: m[3] };
      out.exception = out.exception || {};
      if (!out.exception.address) out.exception.address = m[1];
      return;
    }
    const bare = line.match(/^\[\s*0\]\s+(0x[0-9A-Fa-f]+)\s+(\S+?)\s+\+\s+(0x[0-9A-Fa-f]+)\s+<faulting RIP>/);
    if (bare) {
      out.faultingModule = out.faultingModule || { name: bare[2], rva: bare[3] };
      out.exception = out.exception || {};
      if (!out.exception.address) out.exception.address = bare[1];
      return;
    }
  }
}

function parseRedScriptCallStack(lines, out) {
  out.redScriptStack = [];
  for (const line of lines) {
    const m = line.match(/^\[\s*\d+\]\s+(\S.*?)\s+\(([+\-]?[\d.]+)s\s+ago\)/);
    if (m) out.redScriptStack.push(trim(m[1]));
  }
}

function parseLastActive(lines, out) {
  out.lastActive = out.lastActive || {};
  out.lastActive.scriptedThreads = [];
  out.lastActive.modCodeOnStack = [];
  out.lastActive.liveGameState = {};
  let mode = null;
  for (const line of lines) {
    const sub = trim(line);
    if (/^Scripted code/.test(sub)) { mode = 'scripted'; continue; }
    if (/^Live game state/.test(sub)) { mode = 'gameState'; continue; }
    if (/^Mod code on native stack/.test(sub)) { mode = 'modcode'; continue; }
    if (/^Mods changed since last launch/.test(sub)) { mode = 'mods'; continue; }
    if (mode === 'scripted') {
      const m = line.match(/^\s*(\[CRASHING\])?\s*tid=(\d+)\s+(\S.*?)\s+\(\s*([+\-]?[\d.]+)s\s+ago\)/);
      if (m) {
        out.lastActive.scriptedThreads.push({
          isCrashing: m[1] != null && m[1] !== '',
          tid: parseInt(m[2], 10),
          name: trim(m[3]),
          secondsAgo: parseFloat(m[4]),
        });
      }
    } else if (mode === 'gameState') {
      const m = line.match(/^\s+(\S+)\s+(.+?)\s*$/);
      if (m) out.lastActive.liveGameState[m[1]] = trim(m[2]);
    } else if (mode === 'modcode') {
      const m = line.match(/^\s*\[\s*(\d+)\]\s+(0x[0-9A-Fa-f]+)\s+(\S+)\s+\+\s+(0x[0-9A-Fa-f]+)/);
      if (m) out.lastActive.modCodeOnStack.push({
        frameIdx: parseInt(m[1], 10), pc: m[2], module: m[3], offset: m[4],
      });
    }
  }
}

function parseInstalledModsSummary(lines, out) {
  for (const line of lines) {
    const mgr = line.match(/^Manager:\s*(\S+)\s*(.*)$/);
    if (mgr) {
      out.modManager = mgr[1];
      const rest = trim(mgr[2]);
      out.modManagerDetails = rest ? `${mgr[1]} ${rest}` : mgr[1];
    }
    const counts = line.match(/^(\d+)\s+installed,\s+(\d+)\s+enabled/);
    if (counts) {
      out.installedSummary = {
        installed: parseInt(counts[1], 10),
        enabled: parseInt(counts[2], 10),
      };
    }
  }
}

function parseEngineStateInline(lines, out) {
  out.engineState = out.engineState || {};
  for (const line of lines) {
    const m = line.match(/^(.+?)\s*:\s*(.*)$/);
    if (!m) continue;
    const k = trim(m[1]);
    const v = trim(m[2]);
    if (k === 'Engine state') out.engineState.state = v;
    else if (k === 'Scripts loaded') out.engineState.scriptsLoaded = v === 'true';
    else if (k === 'Engine closing') out.engineState.closing = v.slice(0, 4) === 'true';
    else if (k === 'OOM suspected') out.engineState.oomSuspected = v.slice(0, 4) === 'true';
    else if (k === 'GPU memory') {
      const g = v.match(/(\d+)\s*\/\s*(\d+)\s*MB used\s*\(budget\s+(\d+)/);
      if (g) {
        out.engineState.gpuUsedMB = parseInt(g[1], 10);
        out.engineState.gpuDedicatedMB = parseInt(g[2], 10);
        out.engineState.gpuBudgetMB = parseInt(g[3], 10);
      }
    }
  }
}

function buildStackModules(out) {
  out.stackModules = [];
  const counts = {};
  const order = [];
  for (const e of (out.lastActive && out.lastActive.modCodeOnStack) || []) {
    if (!e.module) continue;
    if (counts[e.module] == null) { counts[e.module] = 0; order.push(e.module); }
    counts[e.module] += 1;
  }
  for (const m of order) {
    out.stackModules.push({ name: m, hits: counts[m], kind: classifyStackKind(m) });
  }
}

function buildScriptedFrames(out) {
  out.scriptedFrames = [];
  if (out.redScriptStack && out.redScriptStack.length > 0) {
    for (let i = out.redScriptStack.length - 1; i >= 0; --i) out.scriptedFrames.push(out.redScriptStack[i]);
    return;
  }
  for (const e of (out.lastActive && out.lastActive.scriptedThreads) || []) {
    if (e.isCrashing && e.name) out.scriptedFrames.push(e.name);
  }
}

function buildGameStateLive(out) {
  if (out.lastActive && out.lastActive.liveGameState) {
    const g = out.lastActive.liveGameState;
    let any = false;
    for (const _ in g) { any = true; break; }
    if (any) out.gameStateLive = g;
  }
}

function inferFaultSiteUnreliable(out) {
  out.fingerprint = out.fingerprint || {};
  out.fingerprint.faultSiteUnreliable =
    !!(out.likelyCulprit && out.likelyCulprit.indexOf('does not lie inside any loaded module') !== -1);
}

function parseCrash(text) {
  const out = { schema: 3, source: 'crash-text-fallback', warnings: [] };
  if (!text) { out.warnings.push('empty input'); return out; }
  const lines = splitLines(text);
  const { sections, order } = splitSections(lines);
  parseHeader(sections.header || [], out);
  if (sections['Crash']) parseCrashSection(sections['Crash'], out);
  for (const label of order) {
    if (startsWith(label, 'LAST ACTIVE')) parseLastActive(sections[label], out);
    else if (startsWith(label, 'Engine state')) parseEngineStateInline(sections[label], out);
    else if (startsWith(label, 'Native call stack')) parseNativeCallstack(sections[label], out);
    else if (startsWith(label, 'RedScript call stack')) parseRedScriptCallStack(sections[label], out);
    else if (startsWith(label, 'Installed mods')) parseInstalledModsSummary(sections[label], out);
  }
  buildStackModules(out);
  buildScriptedFrames(out);
  buildGameStateLive(out);
  inferFaultSiteUnreliable(out);
  if (out.crashId == null) out.warnings.push('no crashId extracted');
  if (out.exception == null) out.warnings.push('no exception extracted');
  return out;
}

File: analyzer/core/crashReportParser.lua
local Parser = {}

local function trim(s)
  if s == nil then return "" end
  return tostring(s):match("^%s*(.-)%s*$") or ""
end

local function nilIfEmpty(s)
  if s == nil or s == "" then return nil end
  return s
end

local function splitLines(text)
  local out = {}
  if text == nil then return out end
  for line in tostring(text):gmatch("([^\r\n]*)\r?\n?") do
    out[#out + 1] = line
  end
  if #out > 0 and out[#out] == "" then out[#out] = nil end
  return out
end

local function sectionLabel(line)
  return line:match("^%-%-%-%s+(.-)%s+%-+%s*$")
end

local function startsWith(s, prefix)
  return s:sub(1, #prefix) == prefix
end

local kFrameworkDlls = {
  ["redscope.dll"]=true, ["red4ext.dll"]=true,
  ["cyber_engine_tweaks.asi"]=true, ["scc.dll"]=true,
}

local function classifyStackKind(moduleName)
  if moduleName == nil or moduleName == "" then return "unknown" end
  local n = string.lower(moduleName)
  if n == "cyberpunk2077.exe" then return "game" end
  if kFrameworkDlls[n] then return "framework" end
  local sys = {
    ["ntdll.dll"]=true, ["kernel32.dll"]=true, ["kernelbase.dll"]=true,
    ["user32.dll"]=true, ["ucrtbase.dll"]=true, ["msvcp140.dll"]=true,
    ["vcruntime140.dll"]=true, ["vcruntime140_1.dll"]=true,
    ["gdi32.dll"]=true, ["gdi32full.dll"]=true, ["combase.dll"]=true,
    ["ole32.dll"]=true, ["oleaut32.dll"]=true, ["shell32.dll"]=true,
    ["advapi32.dll"]=true, ["ws2_32.dll"]=true, ["d3d12.dll"]=true,
    ["dxgi.dll"]=true, ["d3d11.dll"]=true, ["dbghelp.dll"]=true,
    ["msvcrt.dll"]=true, ["rpcrt4.dll"]=true, ["sechost.dll"]=true,
    ["bcrypt.dll"]=true, ["bcryptprimitives.dll"]=true, ["win32u.dll"]=true,
  }
  if sys[n] then return "system" end
  return "mod"
end

local function splitSections(lines)
  local sections = { header = {} }
  local order = { "header" }
  local current = "header"
  for _, line in ipairs(lines) do
    local label = sectionLabel(line)
    if label ~= nil then
      current = label
      if sections[current] == nil then
        sections[current] = {}
        order[#order + 1] = current
      end
    else
      table.insert(sections[current], line)
    end
  end
  return sections, order
end

local function parseHeader(lines, out)
  for _, line in ipairs(lines) do
    local culprit = line:match("^LIKELY CULPRIT:%s*(.*)$")
    if culprit then out.likelyCulprit = trim(culprit) end

    local human, loose = line:match("^CRASH ID:%s*(%S+)%s*%(loose%s+(%S+)%)")
    if human then
      out.crashId = trim(human)
      out.looseId = trim(loose)
    end

    local fpStrict, fpLoose, modset = line:match("fingerprint%s+(%x+)%s*/%s*(%x+)%s+modset%s+(%x+)")
    if fpStrict then
      out.fingerprint = { primary = fpStrict, loose = fpLoose, modSet = modset }
    end
  end
end

local function parseCrashSection(lines, out)
  for _, line in ipairs(lines) do
    local excCode, excName = line:match("^Exception%s*:%s*(0x%x+)%s*(.*)$")
    if excCode then
      out.exception = out.exception or {}
      out.exception.code = excCode
      out.exception.name = trim(excName)
    end

    local addr = line:match("^Address%s*:%s*(0x%x+)")
    if addr then
      out.exception = out.exception or {}
      out.exception.address = addr
    end

    local faultOp, faultAddr = line:match("^Fault data:%s+(%S+)%s+at%s+(0x%x+)")
    if faultOp then
      out.faultData = { op = faultOp, address = faultAddr }
    end

    local modName, modOff = line:match("^%s+module:%s+(%S+)%s*%+%s*(0x%x+)")
    if modName then
      out.faultingModule = { name = modName, rva = modOff }
    end

    local timeStr = line:match("^Time%s*:%s*(.*)$")
    if timeStr then out.timeText = trim(timeStr) end

    local className = line:match("^Class%s*:%s*(.*)$")
    if className then out.crashClass = trim(className) end

    local rsv = line:match("^REDscope%s*:%s*(.*)$")
    if rsv then out.redscopeVersion = trim(rsv) end

    local cpv = line:match("^Cyberpunk%s*:%s*(.*)$")
    if cpv then out.buildId = trim(cpv) end
  end
end

local function parseLastActive(lines, out)
  out.lastActive = out.lastActive or {}
  out.lastActive.scriptedThreads = {}
  out.lastActive.modCodeOnStack = {}
  out.lastActive.liveGameState = {}
  local mode = nil
  for _, line in ipairs(lines) do
    local sub = trim(line)
    if sub:match("^Scripted code") then
      mode = "scripted"
    elseif sub:match("^Live game state") then
      mode = "gameState"
    elseif sub:match("^Mod code on native stack") then
      mode = "modcode"
    elseif sub:match("^Mods changed since last launch") then
      mode = "mods"
    elseif mode == "scripted" then
      local isCrashing = true
      local tid, name, secs = line:match("^%s*%[CRASHING%]%s*tid=(%d+)%s+(%S.-)%s+%(%s*([%+%-]?[%d%.]+)s%s+ago%)")
      if tid == nil then
        isCrashing = false
        tid, name, secs = line:match("^%s*tid=(%d+)%s+(%S.-)%s+%(%s*([%+%-]?[%d%.]+)s%s+ago%)")
      end
      if tid and name then
        table.insert(out.lastActive.scriptedThreads, {
          isCrashing = isCrashing,
          tid = tonumber(tid),
          name = trim(name),
          secondsAgo = tonumber(secs),
        })
      end
    elseif mode == "gameState" then
      local k, v = line:match("^%s+(%S+)%s+(.-)%s*$")
      if k and v ~= "" then
        out.lastActive.liveGameState[k] = trim(v)
      end
    elseif mode == "modcode" then
      local idx, pc, mod, off = line:match("^%s*%[%s*(%d+)%]%s+(0x%x+)%s+(%S+)%s+%+%s+(0x%x+)")
      if idx then
        table.insert(out.lastActive.modCodeOnStack, {
          frameIdx = tonumber(idx),
          pc = pc,
          module = mod,
          offset = off,
        })
      end
    end
  end
end

local function parseNativeCallstack(lines, out)
  for _, line in ipairs(lines) do
    local addr, mod, rva = line:match("^%[%s*0%]%s+(0x%x+)%s+(%S-)!.-%+(0x%x+)%s+<faulting RIP>")
    if addr == nil then
      addr, mod, rva = line:match("^%[%s*0%]%s+(0x%x+)%s+(%S-)%s+%+%s+(0x%x+)%s+<faulting RIP>")
    end
    if addr then
      out.faultingModule = out.faultingModule or { name = mod, rva = rva }
      out.exception = out.exception or {}
      if out.exception.address == nil then out.exception.address = addr end
      return
    end
  end
end

local function parseRedScriptCallStack(lines, out)
  out.redScriptStack = {}
  for _, line in ipairs(lines) do
    local name = line:match("^%[%s*%d+%]%s+(%S.-)%s+%(([%+%-]?[%d%.]+)s%s+ago%)")
    if name then table.insert(out.redScriptStack, trim(name)) end
  end
end

local function parseInstalledModsSummary(lines, out)
  for _, line in ipairs(lines) do
    local mgr, rest = line:match("^Manager:%s*(%S+)%s*(.*)$")
    if mgr then
      out.modManager = mgr
      rest = trim(rest)
      if rest ~= "" then
        out.modManagerDetails = mgr .. " " .. rest
      else
        out.modManagerDetails = mgr
      end
    end
    local installed, enabled = line:match("^(%d+)%s+installed,%s+(%d+)%s+enabled")
    if installed then
      out.installedSummary = {
        installed = tonumber(installed),
        enabled = tonumber(enabled),
      }
    end
  end
end

local function parseBreadcrumbs(lines, out)
  out.breadcrumbs = {}
  for _, line in ipairs(lines) do
    local secs, tag, msg = line:match("^%[%s*([%+%-][%d%.]+)s%]%s+%[([^%]]*)%]%s+(.*)$")
    if secs then
      table.insert(out.breadcrumbs, {
        secondsRelToCrash = tonumber(secs),
        tag = trim(tag),
        message = msg,
      })
    end
  end
end

local function parseRegisters(lines, out)
  out.registers = {}
  for _, line in ipairs(lines) do
    local reg, val, annot = line:match("^(R%w%w?)%s+(0x%x+)%s*(.*)$")
    if reg then
      out.registers[trim(reg)] = { value = val, annotation = trim(annot) }
    end
  end
end

local function parseEngineStateInline(lines, out)
  out.engineState = out.engineState or {}
  for _, line in ipairs(lines) do
    local k, v = line:match("^(.-)%s*:%s*(.*)$")
    if k then
      k = trim(k)
      v = trim(v)
      if k == "Engine state" then out.engineState.state = v
      elseif k == "Scripts loaded" then out.engineState.scriptsLoaded = (v == "true")
      elseif k == "Engine closing" then out.engineState.closing = (v:sub(1, 4) == "true")
      elseif k == "OOM suspected" then out.engineState.oomSuspected = (v:sub(1, 4) == "true")
      elseif k == "GPU memory" then
        local used, ded, budget = v:match("(%d+)%s*/%s*(%d+)%s*MB used%s*%(budget%s+(%d+)")
        if used then
          out.engineState.gpuUsedMB = tonumber(used)
          out.engineState.gpuDedicatedMB = tonumber(ded)
          out.engineState.gpuBudgetMB = tonumber(budget)
        end
      end
    end
  end
end

local function parseEngineStateSibling(text, out)
  if text == nil or text == "" then return end
  out.engineState = out.engineState or {}
  for _, line in ipairs(splitLines(text)) do
    local k, v = line:match("^(.-)%s*:%s*(.*)$")
    if k then
      k = trim(k)
      v = trim(v)
      if k == "Loading stage" then out.engineState.loadingStage = v
      elseif k == "World" then out.engineState.world = v
      elseif k == "Is in game" then out.engineState.isInGame = (v == "true")
      elseif k == "uptimeSeconds" then out.engineState.uptimeSeconds = tonumber(v)
      end
    end
  end
end

local function buildStackModules(out)
  out.stackModules = {}
  local counts = {}
  local order = {}
  for _, entry in ipairs(out.lastActive and out.lastActive.modCodeOnStack or {}) do
    local mod = entry.module
    if mod and mod ~= "" then
      if counts[mod] == nil then
        counts[mod] = 0
        order[#order + 1] = mod
      end
      counts[mod] = counts[mod] + 1
    end
  end
  for _, mod in ipairs(order) do
    table.insert(out.stackModules, {
      name = mod,
      hits = counts[mod],
      kind = classifyStackKind(mod),
    })
  end
end

local function buildScriptedFrames(out)
  out.scriptedFrames = {}
  if out.redScriptStack and #out.redScriptStack > 0 then
    for i = #out.redScriptStack, 1, -1 do
      table.insert(out.scriptedFrames, out.redScriptStack[i])
    end
    return
  end
  for _, entry in ipairs(out.lastActive and out.lastActive.scriptedThreads or {}) do
    if entry.isCrashing and entry.name and entry.name ~= "" then
      table.insert(out.scriptedFrames, entry.name)
    end
  end
end

local function buildGameStateLive(out)
  if out.lastActive and out.lastActive.liveGameState then
    local any = false
    for _ in pairs(out.lastActive.liveGameState) do any = true break end
    if any then out.gameStateLive = out.lastActive.liveGameState end
  end
end

local function inferFaultSiteUnreliable(out)
  out.fingerprint = out.fingerprint or {}
  if out.likelyCulprit and out.likelyCulprit:find("does not lie inside any loaded module", 1, true) then
    out.fingerprint.faultSiteUnreliable = true
  else
    out.fingerprint.faultSiteUnreliable = false
  end
end

function Parser.parseCrash(text)
  local out = {
    schema = 3,
    source = "crash-text-fallback",
    warnings = {},
  }
  if text == nil or text == "" then
    table.insert(out.warnings, "empty input")
    return out
  end

  local lines = splitLines(text)
  local sections, order = splitSections(lines)

  parseHeader(sections.header or {}, out)
  if sections["Crash"] then parseCrashSection(sections["Crash"], out) end

  for _, label in ipairs(order) do
    if startsWith(label, "LAST ACTIVE") then
      parseLastActive(sections[label], out)
    elseif startsWith(label, "Breadcrumbs") then
      parseBreadcrumbs(sections[label], out)
    elseif label == "Registers" then
      parseRegisters(sections[label], out)
    elseif startsWith(label, "Engine state") then
      parseEngineStateInline(sections[label], out)
    elseif startsWith(label, "Native call stack") then
      parseNativeCallstack(sections[label], out)
    elseif startsWith(label, "RedScript call stack") then
      parseRedScriptCallStack(sections[label], out)
    elseif startsWith(label, "Installed mods") then
      parseInstalledModsSummary(sections[label], out)
    end
  end

  buildStackModules(out)
  buildScriptedFrames(out)
  buildGameStateLive(out)
  inferFaultSiteUnreliable(out)

  if out.crashId == nil then table.insert(out.warnings, "no crashId extracted") end
  if out.exception == nil then table.insert(out.warnings, "no exception extracted") end

  return out
end

function Parser.parseFromCombined(text)
  if text == nil or text == "" then return nil end
  local crashText = text
  local stateText = nil
  local marker = "\n--- engine-state.txt (sibling) ---\n"
  local pos = text:find(marker, 1, true)
  if pos then
    crashText = text:sub(1, pos - 1)
    stateText = text:sub(pos + #marker)
  end
  local out = Parser.parseCrash(crashText)
  if stateText then parseEngineStateSibling(stateText, out) end
  return out
end

return Parser
